### PR TITLE
[BEAM-1017] Add RedisIO

### PR DESCRIPTION
**Pull request by jbonofre**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
First round providing RedisPubSubIO (unbounded source).

TODO:
* add RedisIO (bounded) to deal with Redis commands and KV
* find a way to embed a Redis server for the tests (not easy as it's not in Java and so architecture specific). For now, I'm using a Redis server on a server to execute the tests.

**Comment by jkff** [comment]
Note to myself to take a look when I have time
R: @jkff

**Comment by ualtinok** [comment]
Hi,

I'm currently using this and I've added PFADD (HyperLogLog) and withExpiration. I want to also add redis pipelining so that a windowed PCollection will output in single transaction. 

Can you guide me through how I can send a merge request for those or as especially adding commands and expiration is pretty simple can you add those?

**Comment by davorbonaci** [comment]
Hi guys -- just checking -- any updates perhaps?

**Comment by jbonofre** [comment]
Resuming on it with itest and mocked service.

**Comment by jbonofre** [comment]
R: @jkff (for IO itself and `RediService`)
CC: @dhalperi & @davorbonaci (especially for `RedisService` and the way of "fake" a service)
CC: @ssisk (especially for the Integration Tests with `PipelineOptions` and Kubernetes/infra scripts)

**Comment by jkff** [comment]
Thanks! I'll need a few days to go over this. To minimize review rounds, can you do an extra pass yourself and make sure it conforms to PTransform Style Guide (e.g. I noticed some code that was handling data loss by logging a message, which is forbidden by the style guide - there might be more stuff); also stuff like typos and Javadoc comment rendering?

Another thing that would help speed up review and make progress incrementally (I'm particularly concerned that when large PRs like this get a lot of comments, the Github UI becomes unmanageable): consider splitting into separate commits or even PRs: it seems like read part, write part and integration tests for them can all be in separate commits at least.

As for several PRs, that raises a concern where the code may be in a committed intermediate state without ITs (i.e. most likely broken), but I think that can be addressed by iterating on the review via PRs on *your* branch. I.e. make them separate PRs from your redis-io branch to your master branch, and still specify me as reviewer. When we get everything in good shape on your branch, then make one big PR to merge this into beam (which might still have some comments, but not as many).

Let me know if this suggested review process makes sense. It's the first time we'd be doing something like this, but I think it may make things easier compared to how we handled previous PRs of similar size. We can use this PR as a training ground to hash out the process.

**Comment by jbonofre** [comment]
Thanks for the recommandations, that makes perfect sense. I will do a new pass regarding style later today.

**Comment by driverpt** [comment]
Is this PR Dead ?

**Comment by jbonofre** [comment]
No, it's just a work in progress on my github: https://github.com/jbonofre/beam/pull/3

For visibility, I think it's better to keep the PR on Apache Beam and use step by step commits.

I will merge my "local" PR as soon as the review is complete.

**Comment by jbonofre** [comment]
By the way, I splitted this PR in smaller chunk to simplify the review. The first part is the Redis store read: https://github.com/jbonofre/beam/pull/3

I'm updating right now to test with an embedded Redis instance.

**Comment by davorbonaci** [comment]
Any updates perhaps?

**Comment by jbonofre** [comment]
Yes work is in progress on a branch on my GitHub. Review in progress with Eugene.

**Comment by jbonofre** [comment]
@jkff Updated PR with local branches.

File: sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        pipeline.multi();
+        batchCount = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext processContext) {
+        KV<String, String> record = processContext.element();
+        pipeline.append(record.getKey(), record.getValue());
+
+        if (batchCount >= DEFAULT_BATCH_SIZE) {
```
Comment (sivu) [inline sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java]: batchCount is not incremented anywhere, nor reset back to 0 in flush()

Comment (jbonofre) [reply sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java]: Good catch, thanks !

File: sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+ * key/value pairs as {@code KV<String, String>}.
+ *
+ * <p>To configure a Redis source, you have to provide Redis server hostname and port number.
+ * Optionally, you can provide a key pattern (to filter the keys). The following example
+ * illustrates how to configure a source:
+ *
+ * <pre>{@code
+ *
+ *  pipeline.apply(RedisIO.read()
+ *    .withConnectionConfiguration(
```
Comment (jkff) [inline sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java]: I think I asked for this in one of the previous reviews but I'm not sure. This looks very verbose. Add a shorthand read().withEndpoint(host, port) and likewise write().withEndpoint(host, port)?

File: sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+          .apply(View.<KV<String, String>>asIterable());
+      PCollection<KV<String, String>> materialized = input
+          .apply("Identity", ParDo.of(new DoFn<KV<String, String>, KV<String, String>>() {
+            @ProcessElement
+            public void processElement(ProcessContext context) {
+              context.output(context.element());
+            }
+      }).withSideInputs(empty));
+      return materialized
+          .apply("Pair with random key", ParDo.of(
```
Comment (jkff) [inline sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java]: Use Reshuffle.viaRandomKey() here instead - JdbcIO has already been updated.

File: sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    PCollection<KV<String,  String>> readNotMatch = readPipeline.apply("ReadNotMatch",
+        RedisIO.read().withConnectionConfiguration(connection).withKeyPattern("foobar*"));
+    PAssert.thatSingleton(readNotMatch.apply(Count.<KV<String, String>>globally())).isEqualTo(0L);
+
+    readPipeline.run();
+  }
+
+  private RedisConnectionConfiguration createConnection(EmbeddedRedis embeddedRedis) {
+    return RedisConnectionConfiguration.create()
+        .withHost("localhost")
```
Comment (jkff) [inline sdks/java/io/redis/src/test/java/org/apache/beam/sdk/io/redis/RedisIOTest.java]: Remind me whether this works over ipv6? Should it be `::1`?

**Comment by jbonofre** [comment]
@jkff I addressed your comments, thanks !

**Comment by jkff** [comment]
LGTM. Test failure is in my code, I'll address it separately.

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
Jenkins builds fails on the Python SDK, not related to this PR.

@jkff do I have your green light to merge this IO anyway ?

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
@jkff do you think I can merge even if Jenkins python build is failing (it's not related directly to this PR) ?

**Comment by jkff** [comment]
Yes, feel free to merge.

**Comment by jbonofre** [comment]
Thanks !

**Pull request merged**